### ISSUE TITLE
Dot notation for dealing with maps

### ISSUE THREAD

**Issue by tjbenton** [sass/sass#1349]
I would like to see the ability to use dot notation to call variables from a map. While the new `map-get()` in 3.3+ is awesome however it can get to be unreadable real quick. 

For example say that you had this map. 

```
$buttons: (
 color: #fff,
 primary: (
  background: #fff,
  color: #000
  ...
 ),
 secondary: (
  background: #000,
  ...
 )
);
```

Currently to call that you can use `map-get()` like this.

```
.button{
 background: map-get($buttons, background);
 color: map-get($buttons, color);
 &-primary{
  background: map-get(map-get($buttons, primary), background);
  &:hover{
   background: lighten(map-get(map-get($buttons, primary), background), 10%);
  }
 }
...
}
```

This example isn't to bad but when you start calling functions inside of functions inside of functions inside of functions things get out of hand quickly and become really hard to follow. Especially if you are adding and subtracting things to get elements to line up correctly

```
top: em(nth-val(map-get($messaging, padding), 1) - map-get($messaging, line-height), map-get($messaging, font-size)) * -1;
```

If we had dot notation for calling variables It would make things like this more readable. 

```
top: em(nth-val($messaging.padding, 1) - $messaging.line-height, $messaging.font-size) * -1
```

This comes in to play even more if you have a map that is more than 2 levels deep.

```
$example-map: (
  foo: (
    baz: (
      qux: (
        quux: (
          garply: (
            waldo: "Hello world"
          )
        )
      )
    ) 
  ) 
);
```

Currently to get `waldo` in the `$example-map` you would do this 

```
content: map-get(map-get(map-get(map-get(map-get(map-get($example-map, foo), baz), qux), quux), garply), waldo);
```

Using dot notation you could get it like this.

```
content: $example-map.foo.baz.qux.quux.garply.waldo;
```

I realize this example is a little excessive but I hope it shows how useful it could be. 

You shouldn't have to worry about old sass code not working with this change since you can't currently have a `.` in a variable name. 

You would still be able to use interpolation with the variables.

```
@mixin button-theme($theme){
 background: $buttons.#{$theme}.background;
 color: $buttons.#{$theme}.color;
}
```

You should still have the ability to combine the dot notation with the other awesome map functions.

```
$map-to-merge: (
 fred: "",
 plugh: "",
 xyzzy: "",
 thud: ""
);
 $temp: map-merge($example-map.foo.bar, $map-to-merge)
```

**Comment by mistersender** [comment]
i agree, this would be very handy!

**Comment by chriseppstein** [comment]
We considered this, as well as bracket notation (E.g. `$map[$key]`). But we are reticent to introduce new syntax and break future compatibility for this case.

Here's how you can write your own `nested-map-get`. Maybe it's something we can include in a future release if people find it useful.

``` scss
@function list-slice($list, $start, $end: length($list)) {
  @if $start > length($list) {
    @return ();
  }
  $new-list: ();
  @for $i from $start through $end {
     $new-list: append($new-list, nth($list, $i), list-separator($list));
  }
  @return $new-list;
}
@function nested-map-get($map, $keys...) {
  @return if(length($keys) == 1,
             map-get($map, nth($keys, 1)),
             nested-map-get(map-get($map, nth($keys, 1)), list-slice($keys, 2)...));
}

$example-map: (
  foo: (
    baz: (
      qux: (
        quux: (
          garply: (
            waldo: "Hello world"
          )
        )
      )
    ) 
  ) 
);
.result {
  deep-access: nested-map-get($example-map, foo, baz, qux, quux, garply, waldo);
}
```

**Comment by Undistraction** [comment]
I think it's a real shame to close this issue. Dot or bracket notation is so much more readable. As people's use of Sass grows in complexity these things really matter.

```
.result {
  deep-access: nested-map-get($example-map, foo, baz, qux, quux, garply, waldo);
}
```

Vs

```
.result {
  deep-access: $example-map.foo.baz.qux.quux.garply.waldo;
}
```

**Comment by mimshwright** [comment]
agreed. please reconsider!

**Comment by ArmorDarks** [comment]
Though, very little chances, :+1: for reconsideration.

@chriseppstein

> But we are reticent to introduce new syntax and break future compatibility for this case.

May we ask, what exactly potentially it may break in future?

Brackets and dot notations are so common in popular programming languages those days that I really hardly imagine what else for they can be used in future.

**Comment by johnnyshields** [comment]
:+1: this would be nice!

**Comment by tjbenton** [comment]
Sorry in advance for the lengthy post, but someone need to do it. This is a reply to [#1739 (comment)](https://github.com/sass/sass/issues/1739#issuecomment-111102236).
@HugoGiraudel, @davidkpiano, @ArmorDarks I understand that sass is a function based language and that's part of the reason I love using it because it makes doing difficult tasks easy; however using functions **alone** to handle maps and lists can make what should be simple tasks seem overly complicated. In my issue #1349 I suggested using dot/bracket notation for variables only, and the way I suggested it to be implemented would not interfere with the sass ecosystem and instead be a huge improvement to the best preprocessor available. The longer implementing dot/bracket notation is ignored, the bigger the issue becomes. There are several notable libraries out there that have implemented their own way of handling maps, including some that @HugoGiraudel has written. The mere fact that there have been so many libraries attempting to solve this problem should be a clear indicator that it’s an issue worth standardizing.  It also means everyone learns how to handle large sass maps in a different way, and therefore every time someone moves jobs they have to change the way they deal with sass maps and that means relearning something that should have been implemented from the start. These various implementations make it harder to understand what's going on in people’s code, and it make it even more difficult for new people who are trying to learn sass to understand what they are seeing.

@davidkpiano Is 100% correct `map-get($map, foo.bar)` is a very bad idea. That's why I didn't suggest using dot notation in this way. However, if you implement dot notation similarly to how other notable languages use dot notation, it would solve these ridiculous problems everyone runs in to with maps.

In `js` and other languages that use dot/bracket notation you **can't** add `.` into the variable/key name for good reason, and it's pretty much the same point that @davidkpiano is trying to make. The difference being sass **does** follow suit and doesn't accept `.` in a variable/key name. If you tried to set a map key to be `foo.bar` you will get an error. The reason it returns an error is because sass interprets it as a css selector and since a css selector can't be followed by `:` an error is returned. Variables can't have a dot in them either for the same reason. So how does his example work? Well a sass key can be a type of `string`, `list`, `number`, `color`, or `map`. This means we can set a key to anything we want, and that allows us to do things like he's showing in his example. The reason his example works is because `nth(&, 1)` is a type of `list`. So when `map-get($map, nth(&, 1))` is called it's looking for key that matches the list of `nth(&, 1)`. If you absolutely have to have a `.` in your key name then you can do the same thing other languages do; just add quotes around the declaration to note that it's a string `"foo.bar"`— and that makes the key valid. 

Here's @davidkpiano example explained in depth, and the live example is on [sassmiester](http://sassmeister.com/gist/61a28cfb00edde026269) if you would like to see the output

``` sass
// `.qux` was added to make it clear that `&` is a list of lists of strings
.foo.bar, .qux{
  $selector: &;
  selector: $selector; // => ((foo.bar,), (.qux,))
  type-of-selector: type-of($selector); // => list

  $first-selector: nth($selector, 1);
  first-selector: $first-selector; // => (foo.bar,)
  first-selector_type: type-of($first-selector); // => list

  $map: (
   $first-selector: "baz",
   foo: (
    bar: "qux"
   ),
   ("foo",): (
    bar: "waldo"
   ),
   nth($first-selector, 1): "garply" // => ".foo.bar": garply
  );


  $keys: map-keys($map);
  keys: $keys; // => ((.foo.bar .qux,), foo, .foo.bar)

  key-1: nth($keys, 1); // => (.foo.bar, .qux)
  key-1_type: type-of(nth($keys, 1)); // => list

  key-2: nth($keys, 2); // => "foo"
  key-2_type: type-of(nth($keys, 2)); // => string

  key-3: nth($keys, 3); // => ".foo.bar"
  key-3_type: type-of(nth($keys, 3)); // => string

  // getting the value of `$first-selector` in map
  get-test_selector: map-get($map, $first-selector); // => "baz"

  // getting the value of key that is a list
  get-test_list: inspect(map-get($map, ("foo",))); // => (bar: waldo)

  // getting a the value of a key when it contains a `.` in the name
  get-test_string-with-dot: map-get($map, ".foo.bar"); // => "garply"
  get-test_unquote-string-with-dot: map-get($map, unquote(".foo.bar")); // => "garply" still because even if you unquote a `string` it's still a type of `string`
}
```

##### These use cases can still use dot/bracket notation without any special functions being added

``` sass
get-test_selector: $map[nth(&, 1)]; // => "baz"
get-test_list: $map[("foo",)] // => (bar: waldo)
get-test_string-with-dot: $map[".foo.bar"]; // => "garply"
get-test_unquote-string-with-dot: $map[unquote(".foo.bar")]; // => "garply" still because even if you unquote a `string` it's still a type of `string`
```

Implementing dot/bracket notation is something that needs to happen. It would make sass tremendously more readable, and it would make it easier for beginners with background knowledge of `js` and other languages that use dot/bracket notation easier to dive in to sass because they don't have to re-learn the same concept of structures/maps in a different way than they already know. The reason why I started using sass is because of it's extensibility, simplicity, and ease of understanding. The way maps(aka structures) are currently implemented makes it more complex than it needs to be. While I understand sass maps and use them on a daily basis, trying to teach a developer how to use sass map functions is harder than it needs to be and every time each developer has said "Well why didn't they just use dot notation? That would be so much easier!”. I believe if there was a poll on implementing dot/bracket notation in sass and it was asked on all the major sass resource sites and email lists the majority of sass developers would agree that sass needs dot/bracket notation.

### Examples and use cases

Here are some more examples and use cases, how they are written with the current implementation of maps, and how they could be written using dot/bracket notation:

#### Setting a key value using other values from the same map.

Here is a [live](http://codepen.io/tjbenton21/pen/WvJWBz?editors=110) example that I had to use on codepen.

``` sass
// base map for this example
$promo-bg: (
  arrow-width: 2.5em, // width of the arrow on the background
  colors: (#178399, #853381, #438d1d, #d44029, #ec6c21),
  offset: 10%, // offset from the container
  rotate: 3deg, // how much to rotate the background
  height: 80%, // height of the background
  width: 400% // total width of the background
);
```

###### With map functions

``` sass
$promo-bg: map-merge($promo-bg, (
  origin-fix: ((100% + map-get($promo-bg, offset)) / (map-get($promo-bg, width) / (100% + map-get($promo-bg, offset)))) / 2
));
```

###### With dot/bracket notation

``` sass
$promo-bg.orgin-fix: ((100% + $promo-bg.offset) / ($promo-bg.width / (100% + $promo-bg.offset))) / 2;
```

#### Calling a dynamic key

This would be a use case for mixins and functions to call a key based off an argument that was passed:

``` sass
// base map for this example
$user-messaging: (
 success: (
   base: #3c763d,
   light: #dff0D8
  ),
  info: (
   base: #31708f,
   light: #d9edf7
  ),
  warning: (
   base: #8a6d3b,
   light: #fcf8e3
  ),
  danger: (
   base: #a94442,
   light: #f2dede
  )
);
```

###### With map functions

``` sass
@function user-messaging($type, $shade: base){
 @return map-get(map-get($user-messaging, $type), $shade);
}
```

###### With dot/bracket notation

``` sass
@function user-messaging($type, $shade: base){
 @return $user-messaging[$type][$shade];
}
```

#### Getting a variable from a deeply nested map

This example is taken from the original post

###### With map functions

``` sass
map-get(map-get(map-get(map-get(map-get(map-get($example-map, foo), baz), qux), quux), garply), waldo);
```

###### With dot/bracket notation

``` sass
$example-map.foo.baz.qux.quux.garply.waldo
```

#### Setting a value on a deeply nested key

This example uses the same map from the original post.

###### With functions

The only real possibility to do this is to use a third party functions, like the one that @HugoGiraudel wrote and shared on [sitepoint](http://www.sitepoint.com/extra-map-functions-sass/), or the one that was written by @Snugug for [Sass-maps](https://github.com/at-import/Sassy-Maps/blob/0.x.x/sass/sassy-maps/_map-set.scss#L13), or another that was written by @cahnory for [PlastiCSS](https://github.com/cahnory/PlastiCSS/blob/master/function/_map-set.scss), or the one that was written by @pascalduez [gist](https://gist.github.com/pascalduez/14ef79d644483fc262da), these are just a few I came across in a single post by @chriscoyier on [csstricks](https://css-tricks.com/snippets/sass/deep-getset-maps/) and there are so many more out there, this was just from the first link in google.

``` sass
// I can't give a single example because it depends on which third party function you decide use
```

###### With dot/bracket notation

``` sass
// setting a value
$example-map.foo.baz.qux.quux.garply.waldo: "found him";

// replacing an existing map with a new map
$example-map.foo.baz.qux.quux.garply: (
 frank: ""
);

// merging a map
$map-to-merge: (
 fred: "",
 plugh: "",
 xyzzy: "",
 thud: ""
);

$example-map.foo.bar: map-merge($example-map.foo.bar, $map-to-merge);
```

#### Using `map-remove` with nested map

###### With map functions

``` sass
// I can't give a single example because it depends on which third party function you decide use
```

###### With dot/bracket notation

``` sass
// I would like to see it like this because it still makes sense, it's less that you have to type out which can reduce complexity
$example-map: map-remove($example-map.foo.baz.qux.quux, garply);

// I can understand if it was implemented like this because it follows the same way that sass currently works
$example-map.foo.baz.qux.quux: map-remove($example-map.foo.baz.qux.quux, garply);
```

#### Using `map-keys`, and `map-values` on a nested map

###### With map functions

``` sass
map-keys(map-get(map-get(map-get(map-get(map-get($example-map, foo), baz), qux), quux), garply));
map-values(map-get(map-get(map-get(map-get(map-get($example-map, foo), baz), qux), quux), garply));
```

###### With dot/bracket notation

``` sass
map-keys($example-map.foo.baz.qux.quux.garply);
map-values($example-map.foo.baz.qux.quux.garply);
```

#### Using `map-has-key` on a nested map

###### With map functions

``` sass
map-has-key(map-get(map-get(map-get(map-get(map-get($example-map, foo), baz), qux), quux), garply));
```

###### With dot/bracket notation

``` sass
map-has-key($example-map.foo.baz.qux.quux, garply);

// in addition you should be able to do this to check if it's undefined
@if $example-map.foo.baz.qux.quux.garply{
 // execute this if `$example-map.foo.baz.qux.quux.garply` is not undefined and is true
}
```

### Bracket notation for sass lists

The bracket notation should also be applied to sass lists because it's still dealing with the same concept of getting and setting things in variables.

#### Getting/setting an item in a list(aka array)

``` sass
$list: ("baz", "qux", "quux", "garply");
```

###### With list functions

``` sass
// getting
nth($list, 1); // => "baz"
nth($list, length($list)); // => "baz"

// setting
$list: set-nth($list, 1, "waldo"); // => ("waldo", "qux", "quux", "garply")
```

###### With bracket notation

Since sass isn't zero based like js the first item in the array would still be accesed through `1` and the last element would be accesed by the length of the list

``` sass
// getting
$list[1]; // => "baz"
$list[length($list)]; // => "garply"

// setting
$list[1]: "waldo"; // => ("waldo", "qux", "quux", "garply")
```

#### Getting/setting an item in a list(aka array) that is nested in a map

``` sass
$nested-list: (
 foo: (
  foobar: (21, 7, 4),
  bar: (
   baz: ("qux", "quux", "garply")
  )
 )
);
```

###### With map/list functions

``` sass
// getting
nth(map-get(map-get(map-get($nested-list, foo), bar), baz), 2) // => "quux"

// setting
// depends on the third party function you're using, because it's not easily possible with native functions
```

###### With dot/bracket notation

``` sass
// getting
$nested-list.foo.bar.baz[2] // => "quux"

// setting
$nested-list.foo.bar.baz[2]: "waldo";
// => The full map after the item has been set
// $nested-list: (
//  foo: (
//   foobar: (21, 7, 4),
//   bar: (
//    baz: ("qux", "waldo", "garply")
//   )
//  )
// );
```

Dot notation is a very simple concept to understand regardless of your coding background, as it is already very common in several other languages. This would make it easer for other developers to learn sass, and utilize the power of sass maps. Implement it correctly would reduce the amount of wasted time by developers creating their own implementations of the sass map functions, as well as definitively solve a problem that others have already tried to hack their way around.

**Comment by ArmorDarks** [comment]
@tjbenton Thanks for your post! I totally support it

Just to mention, `dot` notation not always means using of dots. Some langauge using a another syntax — for exmple, PHP implements `foo->bar` instead of `foo.bar`.

But dot so far seems to be the most common and the most comfortable syntax.

**Comment by davidkpiano** [comment]
@tjbenton P.S. Sassdash supports the lodash way of delimiting keys with dots: `$value: _get($map, 'foo.bar.baz');` and it checks if `foo.bar.baz` is an existing key first.

I'm still on the fence, though. The native Sass `map-get` works like JavaScript's handling of _maps_, not objects, which makes sense since map keys can be anything, not just strings or numbers. You'd still have to do `Map.prototype.get.apply(...` for ES6 maps, which is even more verbose.

**Comment by tjbenton** [comment]
@davidkpiano I have seen the sassdash implementation of `_get` before and I forgot to add it to the list of the different way's libs handle maps. While it does allow you to use dot notation which is cool. It's still a hackey solution that we shouldn't have to deal with, and something like getting and setting a nested property in a map should be standard across the board for sass but every lib implements it differently, which will causes huge issues when you try to import code snippets from a lib because you have to track down how they implemented maps and update the functions to work with the framework you're using.

**Comment by ArmorDarks** [comment]
I still hope that issue at least will be reopened...

**Comment by tjbenton** [comment]
@ArmorDarks He did say

> "Maybe it's something we can include in a future release if people find it useful."
>  @chriseppstein 

If we can get the word out, and let people know about this issue then maybe it will gain some traction and it might get reopened for consideration.

**Comment by ArmorDarks** [comment]
@tjbenton Sure

@HugoGiraudel I know you had different position regarding dot notation in past. Can you make any comments based on https://github.com/sass/sass/issues/1349#issuecomment-120967912 and https://github.com/sass/sass/issues/1739#issuecomment-111181142? Thanks in advance

**Comment by KittyGiraudel** [comment]
I am afraid I will not have time for this. I'll just give my quick thoughts here, not that anyone care. I always applauded Sass for having a very healthy syntax. Things just look CSS-y, and when it says that it is _an extension of CSS_, I think like it is quite true. CSS heavily relies on functions. And Sass doing otherwise would be very risky for the longevity of their relation. 

When you add dot notation, sooner or later you will add bracket notation. And then, anonymous functions. And then more, and more. And one day, you wake up with a language like Stylus, which is basically JavaScript. Now don't get my wrong, I am a big fan of Stylus, but I don't think it is the right solution for most teams. Sass has a simplicity that makes it easy for anybody. Designers, developers, unexperienced, advanced users...

That's for the opinion-only side of my point. Now, map keys can be anything. Any-fucking-thing. I long wondered why but it turned out to be quite handy in some cases and I now congratule Sass core designers for taking that decision. It was a good choice. Anyway, this is when you take this into account that the whole argument falls apart.

Coming to my point: a string key can contain dots. And that's where the whole dot notation starts to stink. In JavaScript, when you want to access `'bar.baz'` from object `foo`, you can't go `foo.bar.baz`. It doesn't work like this. You have to use brackets: `foo['bar.baz']`. As I said, you cannot really implement dot notation without bracket notation. And this is just an _easy_ scenario. What if you have a list or a map as a key? How exactly do you expect dot notation to work? I'm pretty sure you can come up with some weird syntax (hi `foo.('a', 'b', 'c').('a': 1)`), but would it be a good solution? I hardly think so.

Last, but not least, it is very easy to have a `map-deep-get` function to simplify fetching data in nested maps. It takes a few lines, top:

``` scss
@function map-fetch($map, $keys...) {
  @each $key in $keys {
    $map: map-get($map, $key);

    @if type-of($map) == 'null' {
       @return $map;
    }
  }

  @return $map;
}
```

My 2 cents.

**Comment by mistersender** [comment]
This would be tremendously helpful for our sass implementation as well. we have a lot of maps and, like everyone else, a weird way of getting at them since dot/bracket notation isn't supported yet. it really would make our code a lot clearer. Please support it ASAP!

**Comment by farmok** [comment]
I think this would be very useful for the long term growth of Sass. There will always be the concern of over complicating Sass considering ow accessible it is to designers, but this will happen any way, so why not make it consistent with other languages.

**Comment by ArmorDarks** [comment]
> There will always be the concern of over complicating Sass considering ow accessible it is to designers

As a side note, dot or similar notation will make it much easier for designers. I don't know to which designer `map-get($map, foo, bar)` will be more understandable than `$map.foo.bar`

**Comment by bentonam** [comment]
Agreed that dot/bracket notation needs to be supported.  Please reopen.

**Comment by ds8k** [comment]
I'm with @tjbenton on this, it needs to be supported.

**Comment by mdsither** [comment]
couldnt agree more with @tjbenton  #freedotnotation !

**Comment by alexandreescu** [comment]
With @tjbenton on this. Please support.

**Comment by Camars** [comment]
This would be really useful, please reopen!

**Comment by Happelation** [comment]
@tjbenton knows whats up!

**Comment by tabrindle** [comment]
+1 - I agree - although a nested-map-get function might be a viable alternative for now, this nesting practice is not uncommon, and nested maps would be much more readable if it was a native capability. After all, most programming languages live and die because of readability, especially of complicated data structures. 

If compatibility is a really the concern that closed this ticket, I'm sure having an alternate name for the new function with the nesting capability and eventually deprecating the old map function in favor of it eventually could be a viable solution.

**Comment by InsideTraitor** [comment]
Would be great if you'd reopen this bad boy. Prease leconsidel.

**Comment by chriseppstein** [comment]
First off, I appreciate the passion and effort you've all made to make a case for this feature. I know you feel strongly about it.

> `map-get(map-get(map-get(map-get(map-get(map-get($example-map, foo), baz), qux), quux), garply), waldo)`.

We've said elsewhere in the issue tracker that we're on board with changing map-get to accept a var-arg list of keys. We get that successive `map-get` calls are horrible and we don't recomend that anyone write code like above. If you have that use case, you should be using the trivial work around I provided above to define a nested-map-get until map-get has this change natively. 

> `$example-map.foo.baz.qux.quux.garply.waldo`
> 
> `$example-map[foo][baz][qux][quux][garply][waldo]`

These options are just not a significant enough improvement over `map-get($example-map, foo, baz, qux, quux, garply, waldo)` to warrant introducing new syntax. Additionally, the dot notation is basically a non-starter because it only works with a subset of legal keys.

Furthermore, if we add this for maps, syntax parity would require that we also add bracket access for lists as a short cut for `nth()`. Afterall, the same arguments against map-get apply to nth when working with nested lists. Additionally, since maps are ordered pairs and you can pass them to nth() we would now have an incongruity where the ordered access of maps and lists won't work with the same shorthand syntax.

There are lots of places where Sass is more verbose than our programming language counterparts.  Our general preference is to follow CSS’s lead and to prefer verbosity for the reader over shorthands for the writer. In this case, we think the remedies we have agreed to sufficiently address the core concerns.

**Comment by ArmorDarks** [comment]
After almost half year passed, I'm trying to accept Sass team's arguments so hard, but I just can't.

Sass was supposed to make CSS better and fun again, it supposed to push it further and show the world how modern cascade style sheets have to be done, not follow it's descend into pitfall and 'lets ditch anything that is normal in tons of other languages for no reason' ways...

Let's continue that fest, and replace variables with calls through functions. The only way to call variable through `callSassVariable($variable)` sounds like heresy? To bad for you, because in CSS4 you will have to call variables through functions! Despite it so absurdly for most modern languages.

I don't think that imperfection and сrippleness of CSS ecosystem is good justification for denial of any common today's approach. Otherwise, why would we need Sass at all?

And this is good example, when good intention to keep something simple leads to system, which simple on the paper, but much more complex in practice than should be. For most beginners it's easier to understand concept of dot and brackets notations than get what the hell is `map-set()` and `map-get()`. It so fun to watch newcomers which desperately trying to understand that mess of functions. Possibility to list arguments in `map-get` helps a little, but it still sometimes way more confusing and time-taking than just old good dot notation to set and get something.

I think most part of subscribers silently accepted that decision just because of politeness and respect. But what if we didn't accept it in our hearts? Does it make any difference, when everything decided without us?

I have huge respect for all work done for Sass, I can only imagine how much time it takes daily to maintain it, and I'm deeply thankful for it, I don't want to offend anyone with my post. And I know that sometimes it's better to stick to your own opinion than listen to the crowd. That's true.

But please. Keep Sass fun.

**Comment by davidkpiano** [comment]
@ArmorDarks There are recursive/deep map functions coming up in the next version of Sass. For now, you can easily just use a library like [Sassdash](https://github.com/davidkpiano/sassdash) if you really want dot notation:

``` scss
$palette: (
    'primary': (
        'default': red,
        'dark': darken(red, 10%)
    )
);

$color-dark: _get($palette, 'primary.dark');
// => #CC0000
```

**Comment by KittyGiraudel** [comment]
@davidkpiano The fact that the 2nd argument of the _get method contains dots does not make it a dot accessor. This would be: `$palette.primary.dark`.

I gave my opinion on this topic already: dot notation is fine when you use strings as keys, like JavaScript objects. It is a whole another story when a map key can be of any type. 

```
$map.(0, 1, 2).($foo.25px).baz.null.($bar.false)
```

**Comment by ArmorDarks** [comment]
@davidkpiano This doesn't solve situations, when you have nested lists (arrays) inside maps (objects). You will get same mess of nested functions.

Sassdash makes it better, that true. But making things better not always means solving them. Particularly, Sassdash's `_get` due to slicing and substitutions will have larger impact on performance on large projects (my assumption) than native implementation.

@HugoGiraudel Can't tell for all, but I've never encountered in production Sass maps with keys as anything. And if you will, just for that cases you can use already defined `map-get()`. For rest of cases, it's too verbose and unneeded.

This is the reason why in ES2016 we have objects _and_ maps. Because maps needed in very few situations.

And don't forget bracket notation, which solves that issue too. Though, I assume parenthesis exactly representing bracket notation in your example.

**Comment by mistersender** [comment]
I finally gave up and switched to stylus, dot notation problem: solved!

On Tuesday, May 3, 2016, Serj Lavrin notifications@github.com wrote:

> @davidkpiano https://github.com/davidkpiano This doesn't solve
> situations, when you have nested lists (arrays) inside maps (objects). You
> will get same mess of nested functions.
> 
> Sassdash makes it better, that true. But making things better not always
> means solving them. Particularly, Sassdash's _get due to slicing and
> substitutions will have larger impact on performance on large projects (my
> assumption) than native implementation.
> 
> @HugoGiraudel https://github.com/HugoGiraudel Can't tell for all, but
> I've never encountered in production Sass maps with keys as anything. And
> if you will, just for that cases you can use already defined map-get().
> For rest of cases, it's too verbose and unneeded.
> 
> This is the reason why in ES2016 we have objects _and_ maps. Because maps
> needed in very few situations.
> 
> And don't forget bracket notation, which solves that issue too. Though, I
> assume parenthesis exactly representing bracket notation in your example.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> https://github.com/sass/sass/issues/1349#issuecomment-216538010

## 

Sincerely,

Jessica Kennedy
Triad Webcrafters
www.triadwebcrafters.com
336-298-1511

**Comment by johnnyshields** [comment]
@mistersender thanks for the tip, I'll checkout stylus also!

**Comment by KittyGiraudel** [comment]
@mistersender Of course that is none of my business, but if your choice of CSS preprocessor relies on the ability to use dot-notation, I am afraid you missed the point.

**Comment by linkdesu** [comment]
@michaelklishin THX ! Stylus is really a better choice for node developer !

**Comment by nobuttsseb** [comment]
Over two years later, I guess the feature has not been implemented?

**Comment by ArmorDarks** [comment]
[Modules specs](https://github.com/sass/language/blob/master/proposal/module-system.md) use dot-notation. It gives some hopes.

**Comment by nobuttsseb** [comment]
@ArmorDarks Thanks, I'll check it out!

@pascalduez What's up with the thumbs down? Was my question flawed in some way? Just curious.

**Comment by pascalduez** [comment]
@nobuttsseb sorry, it's an error...

**Comment by ZoltanTompa-ResDiary** [comment]
This is still not implemented in 2019 where it should had been at least 5years ago by now... =S
